KAFKA-9673: Filter and Conditional SMTs

tombentley opened the pull request:
* Add Predicate interface
* Add Filter SMT
* Add the predicate implementations defined in the KIP.
* Create abstraction in ConnectorConfig for configuring Transformations and Connectors with the "alias prefix" mechanism
* Add tests and fix existing tests.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

tombentley committed:
KIP-585/KAFKA-9673: Filter and Conditional SMTs

* Add Predicate interface
* Add Filter SMT
* Add the predicate implementations defined in the KIP.
* Create abstraction in ConnectorConfig for configuring Transformations and Connectors with the "alias prefix" mechanism
* Add tests and fix existing tests.

tombentley committed:
Use Predicates config group

tombentley commented:
Still need to review test coverage, but @kkonstantine, @mimaison, @bbejeck you might want to give it an initial pass.

tombentley commented:
@C0urante you might also want to take a look.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/predicates/RecordIsTombstone.java:
```diff
@@ -0,0 +1,48 @@
+
+    @Override
+    public boolean test(R record) {
+        return record.key() == null;
```
I think we want to check the value instead of the key here?
```suggestion
        return record.value() == null;
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/predicates/RecordIsTombstone.java:
```diff
@@ -0,0 +1,48 @@
+import org.apache.kafka.connect.connector.ConnectRecord;
+
+/**
+ * A predicate which is true for records which are tombstones (i.e. have null key).
```
I think a tombstone is defined as a record with a null value, not a null key:
```suggestion
 * A predicate which is true for records which are tombstones (i.e. have null values).
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Filter.java:
```diff
@@ -0,0 +1,51 @@
+
+    @Override
+    public ConfigDef config() {
+        return null;
```
I think this will fail during validation since [transformations must provide non-null ConfigDefs](https://github.com/apache/kafka/blob/67770072da1bd13762af978faaa278c4039167a2/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java#L355-L363). Might want to instantiate a single empty static `ConfigDef` object for the class and just return that?

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/PredicatedTransformation.java:
```diff
@@ -0,0 +1,67 @@
+
+    @Override
+    public ConfigDef config() {
+        return null;
```
Same comment here as with `Filter`; probably want to return a non-null `ConfigDef` here.

tombentley replied:
This is related to the issue discussed about `configure()`.

We could return an empty `ConfigDef` here, but that would be a lie which could ultimately lead some other error if someone tried to use it with `configure()`.

We can't invent a `ConfigDef` schema for this because the `PredicatedTransformer` would need to know about the `Transformer` is was going to be wrapping, but it can't know that before it's been configured with at least the `Transformer`'s `ConfigDef` and it can't be configured before `config()` has been called. So we have a chicken and egg problem. Something (`ConnectorConfig`) must have some _a priori_ knowledge of either `PredicatedTransformer`'s `ConfigDef`, or know how to configure it without needing to call `config()` at all.

Since `PredicatedTransformer` is a purely internal class which will never be directly exposed to Connect users, we're not obliged to stick to the contract of `config()` and `configure()`. i.e. So  both `PredicatedTransformer.config` and `PredicatedTransformer.configure` can throw when called, since we know no one else can call them and we know `ConnectorConfig` never will.

C0urante replied:
Hmmm... I think there might be some awkwardness here with trying to make `PredicatedTransformer` implement the `Transformation` interface. Could we replace every `Transformation` in the `TransformationChain`'s [transformation list](https://github.com/apache/kafka/blob/de6468ae5915298279e229dc64721e01e7d14fab/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java#L33) with a `PredicatedTransformer` and, if there are no predicates configured for a transform by the user, make the default behavior for the `PredicatedTransformer` class to blindly apply its transformation?

This would solve a few problems:

- No risk of users trying to actually use a `PredicatedTransformer` in a connector config, which they may try to do if we don't add logic to prevent it from being picked up during plugin path scanning on startup and logged as an SMT plugin
- No need to implement methods that aren't used
- One code path instead of two for application of transformations
- More flexibility in instantiation and, possibly, the ability to encapsulate some of the `ConfigDef` generation logic in a separate class from `ConnectorConfig` (haven't looked into the specifics of this yet so may not actually be feasible or that elegant)

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java:
```diff
@@ -257,12 +273,25 @@ public boolean includeRecordDetailsInErrorLog() {
-                transformation.configure(originalsWithPrefix(prefix));
-                transformations.add(transformation);
+                Map<String, Object> configs = originalsWithPrefix(prefix);
+                Object predicateAlias = configs.remove("predicate");
+                Object negate = configs.remove("negate");
```
Do we need to remove these properties here, or can we just read them? Removing might cause issues with SMTs that have config properties with these names; would leaving them in be likely to cause issues as well?

tombentley replied:
The [compatibility section](https://cwiki.apache.org/confluence/display/KAFKA/KIP-585%3A+Filter+and+Conditional+SMTs#KIP-585:FilterandConditionalSMTs-Compatibility,Deprecation,andMigrationPlan) of the KIP says that if a connector already has these configs then they'll be masked by the new implicit configs. If we don't remove them here then we'd be passing the KIP-585 configs to a connector which had it's own semantics for those config keys, which would be incorrect.

C0urante replied:
Hmmm... wish we'd caught that earlier. Seems safer to just leave the properties in, but unless we want to call for a re-vote and an extension on the KIP deadline guess we'll have to keep this as-is.

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java:
```diff
@@ -257,12 +273,25 @@ public boolean includeRecordDetailsInErrorLog() {
+                    Predicate<R> predicate = getClass(predicatePrefix + "type").asSubclass(Predicate.class)
+                            .getDeclaredConstructor().newInstance();
+                    predicate.configure(originalsWithPrefix(predicatePrefix));
+                    transformations.add(new PredicatedTransformation<>(predicate, negate == null ? false : Boolean.parseBoolean(negate.toString()), transformation));
```
Just curious, why directly parse the `negate` property here instead of doing that in `PredicatedTransformation::configure`?

tombentley replied:
A more general question is "Why does `PredicatedTransformation` have a special constructor rather than use `configure(Map<String, ?>)`?" This arises because `PredicatedTransformer` is a bit special. In particular is has to be passed an already configured `Predicate` and a `Transformation`, which you couldn't do with normal `Transformation` (which can only be configured with the types supported by `ConfigDef`).

I guess we could use `configure(Map<String, ?>)`, but then we have to instantiate a Map at this call site only to unpick it again in `PredicatedTransformation`, which feels like more work than just having a constructor. But if you prefer the consistency of using `configure()` I'm happy to do it.

tombentley replied:
Actually, maybe I can make this nicer by changing how ConfigDef instantiates the `PredicateTransformer`, let me see...

tombentley replied:
My idea was to let the `PredicatedTransformer` instantiate its delegate. That could be done, I think but `PredicatedTransformer` cannot instantiate its `Predicate` because the configs for that are under the `predicates.<key>` prefix rather than `transformers.<key>`, so there's not a single `Map` which you could pass to `PredicatedTransformer.configure()` to let it instantiate both transformer and predicate. We could instantiate the predicate in `ConnectorConfig`, but pass a `Map` and let the `PredicatedTransformer` instantiate the `Transformer`, but then transformer instantiation happens in different places depending on whether it's its predicated or not, so it doesn't seem worth it. I'm back to "if you prefer the consistency of using `configure()` I'm happy to do it.", so just let me know.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/predicates/RecordIsTombstone.java:
```diff
@@ -0,0 +1,48 @@
+public class RecordIsTombstone<R extends ConnectRecord<R>> implements Predicate<R> {
+    @Override
+    public ConfigDef config() {
+        return new ConfigDef();
```
Probably won't impact performance too much but we could technically use a single `ConfigDef` instance for the entire class instead of creating a new one every time this method is called.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/predicates/HasHeaderKey.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Override
+    public void configure(Map<String, ?> configs) {
+        this.name = (String) configs.get(NAME_CONFIG_KEY);
```
Might consider using a [SimpleConfig](https://github.com/apache/kafka/blob/67770072da1bd13762af978faaa278c4039167a2/connect/transforms/src/main/java/org/apache/kafka/connect/transforms/util/SimpleConfig.java) here. Won't make a huge difference with the class as-is, but will make it easier to make changes in the future if we ever want to expand on the configurability of this predicate.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/predicates/HasHeaderKey.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Override
+    public boolean test(R record) {
+        return record.headers().allWithName(name).hasNext();
```
Might want to break this up to avoid an NPE since [Headers::allWithName](https://github.com/apache/kafka/blob/67770072da1bd13762af978faaa278c4039167a2/connect/api/src/main/java/org/apache/kafka/connect/header/Headers.java#L52) is technically allowed to return `null` in some situations:
```suggestion
        Iterator<Header> headersWithName = record.headers().allWithName(name);
        return headersWithName != null ? headersWithName.hasNext() : false;
```

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java:
```diff
@@ -257,12 +273,25 @@ public boolean includeRecordDetailsInErrorLog() {
+                if (predicateAlias != null) {
+                    String predicatePrefix = "predicates." + predicateAlias + ".";
+                    @SuppressWarnings("unchecked")
+                    Predicate<R> predicate = getClass(predicatePrefix + "type").asSubclass(Predicate.class)
+                            .getDeclaredConstructor().newInstance();
```
Blegh, was hoping we might be able to use `Utils::newInstance` or `AbstractConfig::newConfiguredInstance` but it looks like neither quite does what we need; the former doesn't do casting to a subclass unless you give it the FQCN of a class instead of an already-loaded `Class<?>` object, and the latter doesn't give enough control over exactly which properties the new instance is configured with.

Since we're using this same logic in several different places in this file alone, we might consider expanding the `Utils` class with a new utility method that does this for us. Maybe something like:

```java
public static Class<T> newInstance(Class<?> klass, Class<T> baseClass) {
    // Return an instance of klass that has been automatically cast to the type of baseClass
}
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/predicates/TopicNameMatches.java:
```diff
@@ -0,0 +1,72 @@
+
+    @Override
+    public void configure(Map<String, ?> configs) {
+        this.pattern = compile(PATTERN_CONFIG_KEY, configs.get(PATTERN_CONFIG_KEY));
```
Same comment as elsewhere: might want to use a `SimpleConfig` in this class.

C0urante reviewed (commented):
Looking pretty good @tombentley! Thanks for the ping.

I was able to get through all of the non-test files except for `ConnectorConfig`; will try to return and make another pass by the end of the week to cover the rest.

Also, do you think it might be worth it to write some integration tests for this stuff? I'm thinking at the very least it'd be cool to verify that `Predicate`s work end-to-end for both source and sink connectors, and since we'd need to pick some SMTs to test the functionality of the `Predicate` plugin with, it might be a convenient excuse to also end-to-end test the new `Filter` SMT introduced here.

tombentley committed:
Some review comments

tombentley committed:
fixup

tombentley committed:
Comment

tombentley committed:
tidy and fix a couple of tests

tombentley commented:
@C0urante thanks for the review, some excellent points there! I think an integration test is a great idea, which I'll work on next. I've addressed all your other comments.

tombentley committed:
Numerous fixes and an integration test.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java:
```diff
@@ -257,12 +274,23 @@ public boolean includeRecordDetailsInErrorLog() {
+                Object negate = configs.remove(PredicatedTransformation.NEGATE_CONFIG);
+                transformation.configure(configs);
+                if (predicateAlias != null) {
+                    String predicatePrefix = "predicates." + predicateAlias + ".";
```
I'd suggest declaring a `PREDICATES_PREFIX` variable for higher visibility.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java:
```diff
@@ -276,116 +304,251 @@ public boolean includeRecordDetailsInErrorLog() {
+                int orderInGroup = 0;
+
+                final String typeConfig = prefix + "type";
+                final ConfigDef.Validator typeValidator = new ConfigDef.Validator() {
```
Consider using `LambdaValidator` to be able to add a lambda for the `toString` method of this validator.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/PredicatedTransformation.java:
```diff
@@ -0,0 +1,81 @@
+
+    static final String PREDICATE_CONFIG = "predicate";
+    static final String NEGATE_CONFIG = "negate";
+    /*test*/ Predicate<R> predicate;
```
this type of comment is not something we use elsewhere and is not immediately obvious what it means. I'd suggest removing instead.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorHandle.java:
```diff
@@ -16,19 +16,21 @@
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 
+import org.apache.kafka.connect.errors.DataException;
```
nit: Our current import style says that these imports stay above the java ones. That's why these show up as changes here.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/TaskHandle.java:
```diff
@@ -16,15 +16,17 @@
+import java.util.function.Consumer;
 import java.util.stream.IntStream;
 
+import org.apache.kafka.connect.errors.DataException;
```
nit: same comment as above

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/TaskHandle.java:
```diff
@@ -37,22 +39,26 @@
     /**
      * Record a message arrival at the task and the connector overall.
      */
-    public void record() {
```
This might break existing tests that depend on Connect's integration tests framework. Probably good idea to keep it, in which case the consumer should be ignored.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ConnectorConfigTest.java:
```diff
@@ -16,21 +16,23 @@
  */
 package org.apache.kafka.connect.runtime;
 
+import java.util.Collections;
```
nit: same as above

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ConnectorConfigTest.java:
```diff
@@ -214,6 +216,187 @@ public void abstractKeyValueTransform() {
         }
     }
 
+
```
nit: 2 extra lines

kkonstantine reviewed (commented):
Thanks for the PR @tombentley. This looks in the right direction. 
I managed to do a first pass. Comments are mostly minor.

Yet, we'll need to consider correct that the new package and interfaces are correctly included or excluded in `PluginUtils` to make these classes compatible with classloading isolation.

tombentley committed:
Review comments

tombentley committed:
Review comments

tombentley commented:
Rebased for conflict.

@kkonstantine I've addressed those first comments, thanks! Still some work on the integration test (not passing when run via gradle).

rhauch commented:
ok to test

rhauch commented on connect/api/src/main/java/org/apache/kafka/connect/transforms/predicates/Predicate.java:
```diff
@@ -0,0 +1,48 @@
+ */
+public interface Predicate<R extends ConnectRecord<R>> extends Configurable, AutoCloseable {
+
+    /**
+     * Configuration specification for this predicate.
+     */
```
```suggestion
    /**
     * Configuration specification for this predicate.
     *
     * @return the configuration definition for this predicate; never null
     */
```

rhauch commented on connect/api/src/main/java/org/apache/kafka/connect/transforms/predicates/Predicate.java:
```diff
@@ -0,0 +1,48 @@
+     */
+    ConfigDef config();
+
+    /**
+     * Returns whether the given record satisfies this predicate.
+     */
```
```suggestion
    /**
     * Returns whether the given record satisfies this predicate.
     *
     * @param record the record to evaluate; may not be null
     * @return true if the predicate matches, or false otherwise
     */
```

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/TransformationIntegrationTest.java:
```diff
@@ -0,0 +1,297 @@
+                .build();
+
+        // start the clusters
+        connect.start();
```
Should we wait until all brokers and Connect workers are available, via something like:
```
        connect.assertions().assertExactlyNumBrokersAreUp(numBrokers, "Brokers did not start in time.");
        connect.assertions().assertExactlyNumWorkersAreUp(numWorkers, "Worker did not start in time.");
```

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/TransformationIntegrationTest.java:
```diff
@@ -0,0 +1,297 @@
+        connectorHandle.expectedCommits(numFooRecords);
+
+        // start a sink connector
+        connect.configureConnector(CONNECTOR_NAME, props);
```
This is an asynchronous method, and it's likely the connector will not be started and running before the test proceeds to the next statements. This can lead to very flaky tests.

We could instead wait until the connector is actually running, using something like:
```
        connect.assertions().assertConnectorAndAtLeastNumTasksAreRunning(CONNECTOR_NAME, NUM_TASKS,
                "Connector tasks did not start in time.");
```

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/predicates/HasHeaderKey.java:
```diff
@@ -0,0 +1,66 @@
+public class HasHeaderKey<R extends ConnectRecord<R>> implements Predicate<R> {
+
+    private static final String NAME_CONFIG = "name";
+    private static final ConfigDef CONFIG_DEF = new ConfigDef().define(NAME_CONFIG, ConfigDef.Type.STRING, null,
```
Here the default is null, which means that the configuration validation allows the `name` field to not be set. Per the KIP, we want to require that `name` is set. To do that, we should use `ConfigDef.NO_DEFAULT_VALUE`:
```suggestion
    private static final ConfigDef CONFIG_DEF = new ConfigDef().define(NAME_CONFIG, ConfigDef.Type.STRING, ConfigDef.NO_DEFAULT_VALUE,
```

tombentley replied:
Ah, thanks, I'd not realised that was the point of `ConfigDef.NO_DEFAULT_VALUE`.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/predicates/TopicNameMatches.java:
```diff
@@ -0,0 +1,75 @@
+public class TopicNameMatches<R extends ConnectRecord<R>> implements Predicate<R> {
+
+    private static final String PATTERN_CONFIG = "pattern";
+    private static final ConfigDef CONFIG_DEF = new ConfigDef().define(PATTERN_CONFIG, ConfigDef.Type.STRING, ".*",
+            new RegexValidator(), ConfigDef.Importance.MEDIUM,
+            "A Java regular expression for matching against the name of a record's topic.");
```
The default doesn't match the KIP. Either we should update the KIP to accept `.*` as the default pattern, or we should use `ConfigDef.NO_DEFAULT_VALUE` as the default to require the `pattern` to be set.

tombentley replied:
I changed it to `.*` only when I realised that the default had to be valid and before I knew about `NO_DEFAULT_VALUE`, so using `NO_DEFAULT_VALUE` is good. Thanks!

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/predicates/TopicNameMatches.java:
```diff
@@ -0,0 +1,75 @@
+    public void configure(Map<String, ?> configs) {
+        SimpleConfig simpleConfig = new SimpleConfig(config(), configs);
+        Pattern result;
+        String value = simpleConfig.getString(PATTERN_CONFIG);
+        try {
+            result = Pattern.compile(value);
+        } catch (PatternSyntaxException e) {
+            throw new ConfigException(PATTERN_CONFIG, value, "entry must be a Java-compatible regular expression: " + e.getMessage());
+        }
```
Can we ever get to line 64? The constructor of the config (line 58) should fail if the `pattern` validator fails to ensure the pattern is a valid regex, which means that if we make it past 58 then line 62 will never fail.

Am I missing something?

rhauch reviewed (commented):
Thanks, @tombentley. I made a first pass, and for the most part this looks good. A few comments/questions below. I'll keep reviewing in more detail.

tombentley committed:
Apply suggestions from code review

Co-authored-by: Randall Hauch <rhauch@gmail.com>

rhauch commented on connect/transforms/src/test/java/org/apache/kafka/connect/transforms/predicates/HasHeaderKeyTest.java:
```diff
@@ -0,0 +1,99 @@
+import static org.junit.Assert.assertTrue;
+
+public class HasHeaderKeyTest {
+
```
Maybe a few more tests that test the ConfigDef and AbstractConfig validation:
```
    @Test
    public void testNameRequiredInConfig() {
        Map<String, String> props = new HashMap<>();
        ConfigException e = assertThrows(ConfigException.class, () -> config(props));
        assertTrue(e.getMessage().contains("Missing required configuration \"name\""));
    }

    @Test
    public void testNameMayNotBeEmptyInConfig() {
        Map<String, String> props = new HashMap<>();
        props.put("name", "");
        ConfigException e = assertThrows(ConfigException.class, () -> config(props));
        assertTrue(e.getMessage().contains("String must be non-empty"));
    }

    protected SimpleConfig config(Map<String, String> props) {
        return new SimpleConfig(new HasHeaderKey().config(), props);
    }
```
BTW, note that the `new HasHeaderKey().config()` is required because there is no accessible static `ConfigDef`. Might want to just make the static field package protected.

tombentley replied:
The empty string is a valid regex, so I used `CompositeValidator.of(new NonEmptyString(), new RegexValidator())` in `TopicNameMatches`, rather than just `new RegexValidator()`.

tombentley replied:
Oops, replied to wrong comment, but I'm sure you guessed what I mean.

rhauch commented on connect/transforms/src/test/java/org/apache/kafka/connect/transforms/predicates/TopicNameMatchesTest.java:
```diff
@@ -0,0 +1,65 @@
+import static org.junit.Assert.assertTrue;
+
+public class TopicNameMatchesTest {
+
```
Maybe a few more tests that test the ConfigDef and AbstractConfig validation:
```
    @Test
    public void testPatternRequiredInConfig() {
        Map<String, String> props = new HashMap<>();
        ConfigException e = assertThrows(ConfigException.class, () -> config(props));
        assertTrue(e.getMessage().contains("Missing required configuration \"pattern\""));
    }

    @Test
    public void testPatternMayNotBeEmptyInConfig() {
        Map<String, String> props = new HashMap<>();
        props.put("pattern", "");
        ConfigException e = assertThrows(ConfigException.class, () -> config(props));
        System.out.println(e.getMessage());
        assertTrue(e.getMessage().contains("String must be non-empty"));
    }

    @Test
    public void testPatternIsValidRegexInConfig() {
        Map<String, String> props = new HashMap<>();
        props.put("pattern", "[");
        ConfigException e = assertThrows(ConfigException.class, () -> config(props));
        System.out.println(e.getMessage());
        assertTrue(e.getMessage().contains("Invalid regex"));
    }

    protected SimpleConfig config(Map<String, String> props) {
        return new SimpleConfig(new TopicNameMatches().config(), props);
    }
```

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginUtilsTest.java:
```diff
@@ -128,6 +131,10 @@ public void testAllowedConnectFrameworkClasses() {
                 "org.apache.kafka.connect.transforms.ExtractField$Key")
         );
+        assertTrue(PluginUtils.shouldLoadInIsolation("org.apache.kafka.connect.transforms.predicates."));
+        assertTrue(PluginUtils.shouldLoadInIsolation(
+                "org.apache.kafka.connect.transforms.predicates.TopicNameMatches")
+        );
```
Should we also check the other two predicate implementations, too? I know it's not strictly required, but it would help to better ensure the pattern matches more than just one implementation.

kkonstantine replied:
Given the small number of predicates, I agree with @rhauch

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ConnectorConfigTest.java:
```diff
@@ -214,6 +216,185 @@ public void abstractKeyValueTransform() {
+        }
+    }
+
+    @Ignore("Is this really an error. There's no actual need for the predicates config (unlike transforms where it defines the order).")
```
It'd be good to have the test reflect the current behavior.

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ConnectorConfigTest.java:
```diff
@@ -214,6 +216,185 @@ public void abstractKeyValueTransform() {
+        props.put("predicates", "my-pred");
+        props.put("predicates.my-pred.type", TestPredicate.class.getName());
+        props.put("predicates.my-pred.int", "79");
+        try {
+            new ConnectorConfig(MOCK_PLUGINS, props);
+            fail();
+        } catch (ConfigException e) {
+            assertTrue(e.getMessage().contains("Value must be at least 80"));
+        }
```
We've moved to using `assertThrows` here, which would look something like:
```suggestion
        ConfigException e = assertThrows(ConfigException.class, () -> new ConnectorConfig(MOCK_PLUGINS, props));
        assertTrue(e.getMessage().contains("Value must be at least 42"));
```

rhauch reviewed (commented):
Got through the test code, and found a few more suggestions.

rhauch commented:
ok to test

rhauch commented:
retest this please

rhauch commented:
ok to test

rhauch committed:
KAFKA-9673: Fix a few unit and integration tests

rhauch committed:
KAFKA-9673: Added some tests, made others a bit more robust, and added more validation to TopicNameMatches

rhauch commented:
Added several commits that fixed the failing unit and integration tests, and addressed several of my other comments.

rhauch commented:
ok to test

rhauch commented:
retest this please

rhauch commented:
retest this please

rhauch commented:
Builds were all aborted: `Aborted by user anonymous`
* https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/2554/
* https://builds.apache.org/job/kafka-pr-jdk11-scala2.13/6574/
* https://builds.apache.org/job/kafka-pr-jdk14-scala2.13/713/

rhauch commented:
retest this please

rhauch reviewed (approved):
Thanks, @tombentley! LGTM, pending a green build (which are currently waiting for executors).

rhauch commented:
ok to test

rhauch commented:
Jenkins jobs were never associated with this PR (see [INFRA-20344](https://issues.apache.org/jira/browse/INFRA-20344)), but here are the jobs for the latest commit:
* https://builds.apache.org/job/kafka-pr-jdk14-scala2.13/715
* https://builds.apache.org/job/kafka-pr-jdk11-scala2.13/6576
* https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/2556

kkonstantine commented:
retest this please

kkonstantine reviewed (approved):
LGTM, pending a green build
Nice work @tombentley !

tombentley committed:
Trivial correction

tombentley commented:
@rhauch @kkonstantine thanks very much for making those fixes for me, I appreciate the time that must've taken. I've made one more trivial correction to the integration test.

rhauch commented:
Thanks, @tombentley. Merging with 3 green builds that Jenkins did not tie back to this PR.
